### PR TITLE
Add fonts to render text in plots.

### PR DESCRIPTION
**Pull request by jdblischak**
Checklist
* [x] Used a fork of the feedstock to propose changes
* [x] Bumped the build number (if the version is unchanged)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

cc: @ocefpaf @bgruening @mingwandroid @dpryan79 

xref: https://github.com/conda-forge/r-base-feedstock/issues/91, https://github.com/conda-forge/r-base-feedstock/pull/104

### CODE REVIEW

**Review by dpryan79 — APPROVED** [review]

**Comment by mingwandroid** [comment]
Can we indirect fonts via a package called `conda-ecosystem-fonts` (or something like that) instead here? I'd like to continue having that point at the current fonts in `anaconda-fonts` for Anaconda Distribution ideally.

Pinging @msarahan and @katietz

File: recipe/meta.yaml [hunk 1]
```diff
@@ -12,6 +12,7 @@ requirements:
   run:
     - r-base
     - r-recommended
+    - mscorefonts
```
Comment (dbast) [inline recipe/meta.yaml:15]: hm...  mscorefonts is only available for linux64 according to https://anaconda.org/conda-forge/mscorefonts ... that would mean that the noarch:generic r package gets uninstallable on mac and windows.

Comment (dbast) [reply recipe/meta.yaml:15]: Or do I overlook something here?

Comment (jdblischak) [reply recipe/meta.yaml:15]: @dbast No, that is a very good point. Thanks for noting it! My main goal has been to fix the text in plots created in Docker containers and other minimal environments. I hadn't thought to check for macOS and Windows builds.

Comment (mingwandroid) [reply recipe/meta.yaml:15]: Fonts should be noarch. Here's the recipe we've been using: https://github.com/AnacondaRecipes/aggregate/blob/master/fonts-anaconda-feedstock/recipe/meta.yaml

The font recipes are also in that repo.

Comment (jdblischak) [reply recipe/meta.yaml:15]: @mingwandroid That recipe looks super useful. Should we migrate it to conda-forge? There has been a lot of effort recently by others to make R only depend on conda-forge packages:

https://github.com/conda-forge/r-base-feedstock/issues/83
https://github.com/conda-forge/r-base-feedstock/issues/64

Comment (mingwandroid) [reply recipe/meta.yaml:15]: Feel free to adopt it. What about metapackage to abstract which endeavour the fonts come from?  So we'd have fonts-anaconda and fonts-conda-forge, and each have different fonts-conda-ecosystem packages to indirect them?

Overall, though, I. do think adding fonts to `r` isn't the right decision for various technical reasons. Only packages that depend upon (at least transitively) our `fontconfig` package will have any possibility of accessing these fonts, so therefore introducing a `run_exports` into the `fontconfig` package is the right thing to do here, then everything just works everywhere automatically. That is very much the reason we came up with `run_exports`.

Comment (mingwandroid) [reply recipe/meta.yaml:15]: We are really saying when we craft the recipe using `run_exports`: "the fontconfig package (whose primary concern is managing fonts on this system) is a dependency of this application, therefore we need to install the datafiles that are part of that overall font configuration".

Comment (msarahan) [reply recipe/meta.yaml:15]: I agree, a metapackage will help this be much more flexible.

Comment (dbast) [reply recipe/meta.yaml:15]: Maybe we soon have a noarch mscorefonts package https://github.com/conda-forge/mscorefonts-feedstock/pull/3 

So the conda-forge version of `fonts-conda-ecosystem` pkg could redirect to mscorefonts (after the PR gets merged) or to `fonts-conda-forge` (after that is created out of fonts-anaconda).

Comment (mingwandroid) [reply recipe/meta.yaml:15]: I would recommend conda-forge adopt (or discuss for adoption) the set I selected and put them all in the metapackage. I was pretty careful to pick fonts that covered a large range of scientific glyphs while keeping size concerns in mind (not that I think size concerns are relevant for our eco-system given we use hardlinks everywhere anyway!).

Comment (jdblischak) [reply recipe/meta.yaml:15]: > I do think adding fonts to r isn't the right decision for various technical reasons.

My goal is that a conda user that installs R inside a Docker container can create plots. I am agnostic to how we implement this.

> Only packages that depend upon (at least transitively) our fontconfig package will have any possibility of accessing these fonts, so therefore introducing a run_exports into the fontconfig package is the right thing to do here, then everything just works everywhere automatically. That is very much the reason we came up with run_exports.

So I need to add `r-base` to the `run_exports` section of the `fontconfig` recipe?

https://github.com/conda-forge/fontconfig-feedstock/blob/master/recipe/meta.yaml#L28

**Comment by mingwandroid** [comment]
> My goal is that a conda user that installs R inside a Docker container can create plots. I am agnostic to how we implement this.

Hey @jdblischak, IMHO that is not the best goal. Implementation details matter for they directly impact maintainability. Maintainability is more important that currently being able to make a plot since that impacts future plotting capabilities.

File: recipe/meta.yaml (hunk 1 above)
Comment (dbast) [reply recipe/meta.yaml:15]: Are you sure? Reading also the other threads, I am seeing the following steps:
1. Creation of the 2 meta-packages and 4 font-packages in conda-forge: `fonts-conda-ecosystem` pointing to `fonts-conda-forge` which depends on `font-ttf-{ubuntu, inconsolata, dejavu-sans-mono, source-code-pro}`.
2. Adding `fonts-conda-ecosystem` to `run_exports` of `fontconfig`: all packages that have `fontconfig` as `host` dependency then receive those fonts as `run` dependency.
3. Updating all font-consuming packages that do not yet depend on fontconfig or any font package (e.g. r-ggplot2): Either adding fontconfig to the `host` dependencies (which then implicitly adds `fonts-conda-ecosystem` to `run` dependencies) or adding directly `fonts-conda-ecosystem` to `run` dependencies... (so no font dependencies are added the `r-base` or `r` package). 
4. Adding some tests scripts (from the other threads) to e.g. r-ggplot2 to proof that the fonts are found and the right types of fonts are available.

Are those the right steps?

Comment (mingwandroid) [reply recipe/meta.yaml:15]: Exactly @dbast. But let us get concensus that this is the best way.

**Comment by mingwandroid** [comment]
An argument could be made for having a further indirection through `fontconfig-with-fonts` which holds a run dep for `fontconfig` and a `run_exports` for `fonts-conda-ecosystem`. My thinking is that other projects that want to depend upon `fontconfig` directly may not agree with our choices of fonts and this gives them an escape route (think packaging an app like `Xpra` with `constructor`).

**Comment by mingwandroid** [comment]
pinging @conda-forge/core since this is/should be ecosystem-wide.

**Comment by dbast** [comment]
Created the 6 mentioned packages https://github.com/conda-forge/staged-recipes/pull/9168 for further discussions.

Maybe it is possible to agree on those packages and later agree on the mechanism to role them out via different ways of `run` dependencies.

**Comment by bgruening** [comment]
How do we proceed here? Do we extend the run dependencies?

**Comment by dbast** [comment]
@bgruening Thanks for bringing this up again.

Could we change this PR to add a run dependency to the meta package `fonts-conda-ecosystem` (instead of `mscorefonts`). That would be more flexible. Any other options on this?

(Independent of this PR: We could also add `fonts-conda-ecosystem` as run dependency to the consumers like r-ggplot2... this is all easy to revert, if the fontconfig step would be done in the future).

**Comment by dpryan79** [comment]
Should we think about implementing this for the R 4.0.0 roll-out?

**Comment by jdblischak** [comment]
Closing this PR. I'm pretty sure this has been fixed https://github.com/conda-forge/r-base-feedstock/issues/91#issuecomment-832352863

**Pull request closed without merging**